logins-sql/ffi/src/lib.rs

```diff
@@ -69,94 +61,85 @@ fn parse_url(url: &str) -> sync15_adapter::Result<url::Url> {
 
 #[no_mangle]
 pub unsafe extern "C" fn sync15_passwords_sync(
-    state: *mut PasswordEngine,
+    state: &mut PasswordEngine,
     key_id: *const c_char,
     access_token: *const c_char,
     sync_key: *const c_char,
     tokenserver_url: *const c_char,
-    error: *mut ExternError
+    error: &mut ExternError
 ) {
     trace!("sync15_passwords_sync");
-    with_translated_void_result(error, || {
-        assert!(!state.is_null(), "Null state passed to sync15_passwords_sync");
-        let state = &mut *state;
+    // TODO: Is there any way to convince rust that some `&mut T` is unwind safe?
+    call_with_result(error, || {
         state.sync(
             &sync15_adapter::Sync15StorageClientInit {
-                key_id: c_str_to_str(key_id).into(),
-                access_token: c_str_to_str(access_token).into(),
-                tokenserver_url: parse_url(c_str_to_str(tokenserver_url))?,
+                key_id: rust_string_from_c(key_id),
+                access_token: rust_string_from_c(access_token),
+                tokenserver_url: parse_url(rust_str_from_c(tokenserver_url))?,
             },
             &sync15_adapter::KeyBundle::from_ksync_base64(
-                c_str_to_str(sync_key).into()
+                rust_str_from_c(sync_key)
             )?
         )
     })
 }
 
 #[no_mangle]
 pub unsafe extern "C" fn sync15_passwords_touch(
-    state: *const PasswordEngine,
+    state: &PasswordEngine,
     id: *const c_char,
-    error: *mut ExternError
+    error: &mut ExternError
 ) {
     trace!("sync15_passwords_touch");
-    with_translated_void_result(error, || {
-        assert!(!state.is_null(), "Null state passed to sync15_passwords_touch");
-        let state = &*state;
-        state.touch(c_str_to_str(id))
+    call_with_result(error, || {
+        state.touch(rust_str_from_c(id))
     })
 }
 
 #[no_mangle]
 pub unsafe extern "C" fn sync15_passwords_delete(
-    state: *const PasswordEngine,
+    state: &PasswordEngine,
     id: *const c_char,
-    error: *mut ExternError
+    error: &mut ExternError
 ) -> u8 {
     trace!("sync15_passwords_delete");
-    with_translated_value_result(error, || {
-        assert!(!state.is_null(), "Null state passed to sync15_passwords_delete");
-        let state = &*state;
-        let deleted = state.delete(c_str_to_str(id))?;
-        Ok(if deleted { 1 } else { 0 })
+    call_with_result(error, || {
+        state.delete(rust_str_from_c(id))
+        .map(|deleted| {
```

Comment:
Didn't you set up machinery to do this automatically?

Reply:
Oh, right, sorry, I added that late and forgot to fix this, thanks.